### PR TITLE
Switch Homeroom table JS to es6 

### PR DESCRIPTION
**Pull request by Jngai**
(cherry picked from commit 93e4323b4c3e6ec9e95d79ad42f1665ace163bec)

# Who is this PR for?
Everybody that takes care of the codebase. Ref #1273 

# What problem does this PR fix?
Split up my work file by file to make it easier for QA to avoid js bugs during the refactoring. Moving work piece by piece from my other PR. 

# What does this PR do?

- Switch `homeroom_table.jsx` to `HomeroomTable.js`. Rewrite it into a js module then import it everywhere it happens. Remove window.shared code. 
- refactored its associated test

# Screenshot (if adding a client-side feature)

# Checklist
+ [x] Author - Do a rigorous code review. 
+ [x] Author - Know where the js is getting loaded, click around
+ [x] Author checked latest in IE - Homeroom Page
+ [x] Reviewer - Do a rigorous code review. 
+ [x] Reviewer checked latest in IE - Homeroom Page

### CODE REVIEW

File: app/assets/javascripts/homeroom_table/HomeroomTable.js [hunk 1]
```diff
@@ -4,34 +4,27 @@ import SortHelpers from '../helpers/sort_helpers.jsx';
 import Cookies from 'js-cookie';
```
Comment (Jngai) [inline app/assets/javascripts/homeroom_table/HomeroomTable.js]: The goal here is to also renames app and test files to look like what @kevinrobinson did in #1224.

**Comment by alexsoble** [comment]
@Jngai Love this approach!! Much more modular and easier to review — thanks!

File: app/assets/javascripts/homeroom_table/HomeroomTable.js [hunk 2]
```diff
@@ -4,36 +4,29 @@ import SortHelpers from '../helpers/sort_helpers.jsx';
...
       showColumnPicker: false,
       sortBy: 'first_name',
       sortType: 'string',
       sortDesc: true
     };
-  },
+  }
 
-  orderedStudents () {
+  orderedStudents() {
```
Comment (Jngai) [inline app/assets/javascripts/homeroom_table/HomeroomTable.js]: @alexsoble This js file has minimum change. The only changes are the `,` and moving in `()` to the function name. Nothing has been removed. All of the functions in here are closed correctly. I double check the `constructor` with `state` getting set.

File: spec/javascripts/homeroom_table/HomeroomTable.test.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import students from '../fixtures/homeroom_students.jsx';
+import ReactDOM from 'react-dom';
+
+const helpers = {
+  renderInto: function(el, props) {
+    ReactDOM.render(<HomeroomTable {...props} />, el);
+  }
+};
+
+SpecSugar.withTestEl('high-level integration test', (container) => {
```
Comment (Jngai) [inline spec/javascripts/homeroom_table/HomeroomTable.test.js]: using `arrow functions` here instead of function definition.

**Comment by alexsoble** [comment]
@Jngai Thanks for adding the review checklist to this PR! I edited it to add Reviewer tasks (tasks for me) as well as Author tasks (tasks for you). As soon as you've checked off your tasks from the list (you can just click the little box in the issue), let me know and I'll do my review tasks. Thanks!

**Comment by Jngai** [comment]
@alexsoble I clicked around a lot in homeroom where the file is loaded. 
screen shot 2017-12-04 at 12 02 43 pm

I fixed a bug and checked off a couple of items. Its time for a review by you. Please review this especially that the syntax for my test refactoring is good too.

**Comment by Jngai** [comment]
@alexsoble please click around in homeroom too. If this is good then I will head on to the next.

**Comment by alexsoble** [comment]
👍 👍

**Comment by alexsoble** [comment]
I checked the header sorting functionality on the homerooms table, and that worked well. 

But when I tried the column toggling functionality, that was broken. I'll add GIFs.

**Comment by alexsoble** [comment]
## Toggle columns (demo site, prod)

select-columns-demo-site

## Toggle columns (testing out this PR locally)

select-columns-pr

**Comment by Jngai** [comment]
Its good now. Will you take a look? Sorry I didn't catch what those 3 dots was for.

recordingtoggle

**Comment by alexsoble** [comment]
@Jngai OK great, thanks! For sure, I would say the meaning of those three dots in the UI isn't the clearest, which is a UI improvement we should make. If anything in the UI is unclear to you, please ask! It's important for you to understand the UI so that you can check your work when you refactor JS. 

The same goes for each function — if you're refactoring a file and you're ever unclear on what a function is doing, please ask. I'll answer or someone else from the team will if they spot it first. It may seem like these ES6 improvements are "mechanical" changes, but they really require you as the developer to understand what every function you're refactoring is doing, so that you can check your work. Never hesitate to ask those kinds of questions!

**Comment by alexsoble** [comment]
@Jngai Often you'll be able to discover the answers for yourself by tracing through the codebase.
 (For example — what code is rendering those three dots? Is there a handler function attached to the click event? What does it do?) But if you run into code that's difficult for you to trace, you can always ask another project member.

**Comment by alexsoble** [comment]
@Jngai We should also talk about test coverage! Let's do what we talked about in #1289. Our CI failed us on this PR, because it passed but the PR would have introduced a bug (or two depending on how you count). Can you add tests that would fail if a user is unable to toggle the column selection menu and then toggle columns on or off?

**Comment by alexsoble** [comment]
@Jngai Also — **much** less important than test coverage, but the `constructor` method on the HomeroomTable was a little difficult for me to read because there were no spaces between any of the lines. I made a small patch for this here: ad2779bf56c489e474da93400564dc687289d5a0, could you cherry-pick it? Goal is to separate out each of the things the constructor is doing with a little bit of space.

**Comment by Jngai** [comment]
@alexsoble I will ask next time. I didn't know those 3 dots was actually a feature...I concluded that it was a decoration on the page. At first I was doing mechanical changes but now I have started to get to know these functions since I need to understand them in order to fix the bugs.

Firefox dev tool will help tremendously here. 

The pr got switch back to WIP so I can add tests to this pr. I wish we can automate more here.

I will definitely cherry pick your commit.

**Comment by alexsoble** [comment]
@Jngai 👍  Sounds good! What kind of things do you want to automate?

**Comment by Jngai** [comment]
@alexsoble I am thinking making automated tests to help us with the clicking so we don't have to do it manually. I see capybara in the app but not many acceptance tests and or any view specs. The jest test whether the component renders or not. I wonder if jest can serve the purpose here but I will have to dive in and see.

**Comment by alexsoble** [comment]
@Jngai I agree, I'm also interested in adding more types of tests to the app. I started exploring integration tests in Internet Explorer using Sauce in #1170, but the authorization refactor work, referential integrity work, and reviewing/checking contributor PRs has taken precedence. I think the for the moment we should bias towards building out our existing jest suite over adding new types of tests, but if you want to do a spike to add a new category of testing down the line that's more than fine with me.

**Comment by Jngai** [comment]
@alexsoble I began doing testing last night. I have a couple of ideas and they all include building out the jest test suite.

**Comment by Jngai** [comment]
@alexsoble I want to use enzyme and see traces of it in the app. Are we not using it? I have two tests that test the toggle columns on and off. Here is the guide that I am looking at: https://medium.com/@newyork.anthonyng/testing-react-components-state-b57bfc712b90 
I think the tests can be better if I directly test if the props change state from true to false and vice versa. I don't see an option exist for this using react test utilities.

**Comment by alexsoble** [comment]
@Jngai I don't see enzyme in package.json, so it looks like we're not using it. To me, it looks like the tests you wrote resemble the "good" tests in the Medium article you shared: they test that clicking component 1 changes component 2, instead of testing the internals of component 1. That's a really good pattern! 

I would say let's add a similar spec for the toggle column functionality itself. Once that's done I think we'll be in good shape to merge.

**Comment by Jngai** [comment]
@alexsoble Ok great, I will work on this and update you soon.

**Comment by Jngai** [comment]
@alexsoble Just added the last test. I can't really cherry pick your spacing commit into my pr so I just added it manually. Its ready for a review. Also I notice that enzyme is a dependency for rc-slider and its in `/node_modules/rc-slider/package.json`.

**Comment by alexsoble** [comment]
Hey @Jngai! Looks good in IE. My only question is about one of the specs:

```
it('able to remove a column when unchecking an item on the column picker menu  ', () => {
  ...
  checkbox.click();
  ...
  expect($(el).find('span.table-header').length).toEqual(7);
});
```

Can we add an expectation to make sure the length was 8 before, so that we can make sure the total decreased by 1 to get to 7?

**Comment by Jngai** [comment]
@alexsoble all set.

**Comment by alexsoble** [comment]
@Jngai Looks good in prod!! Onto the next section!! :)

**Comment by Jngai** [comment]
@alexsoble Great! I will work on the next section.

**Pull request merged**